openstack/networking/v2/extensions/lbaas_v2/l7policies/requests.go

```diff
@@ -0,0 +1,84 @@
+package l7policies
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// CreateOptsBuilder allows extensions to add additional parameters to the
+// Create request.
+type CreateOptsBuilder interface {
+	ToL7PolicyCreateMap() (map[string]interface{}, error)
+}
+
+type Action string
+type RuleType string
+type CompareType string
+
+const (
+	ActionRedirectToPool Action = "REDIRECT_TO_POOL"
+	ActionRedirectToURL  Action = "REDIRECT_TO_URL"
+	ActionReject         Action = "REJECT"
+
+	TypeCookie   RuleType = "COOKIE"
+	TypeFileType RuleType = "FILE_TYPE"
+	TypeHeader   RuleType = "HEADER"
+	TypeHostName RuleType = "HOST_NAME"
+	TypePath     RuleType = "PATH"
+
+	CompareTypeContains  CompareType = "CONTAINS"
+	CompareTypeEndWith   CompareType = "ENDS_WITH"
+	CompareTypeEqual     CompareType = "EQUAL_TO"
+	CompareTypeRegex     CompareType = "REGEX"
+	CompareTypeStartWith CompareType = "STARTS_WITH"
+)
+
+// CreateOpts is the common options struct used in this package's Create
+// operation.
+type CreateOpts struct {
+	// Name of the L7 policy.
+	Name string `json:"name,omitempty"`
+
+	// The ID of the listener.
+	ListenerID string `json:"listener_id" required:"true"`
+
+	// The L7 policy action. One of REDIRECT_TO_POOL, REDIRECT_TO_URL, or REJECT.
+	Action Action `json:"action" required:"true"`
+
+	// The position of this policy on the listener.
+	Position int32 `json:"position,omitempty"`
+
+	// A human-readable description for the resource.
+	Description string `json:"description,omitempty"`
+
+	// TenantID is the UUID of the project who owns the L7 policy in neutron-lbaas.
+	// Only administrative users can specify a project UUID other than their own.
+	TenantID string `json:"tenant_id,omitempty"`
+
+	// ProjectID is the UUID of the project who owns the L7 policy in octavia.
+	// Only administrative users can specify a project UUID other than their own.
+	ProjectID string `json:"project_id,omitempty"`
+
+	// Requests matching this policy will be redirected to the pool with this ID.
+	// Only valid if action is REDIRECT_TO_POOL.
+	RedirectPoolID string `json:"redirect_pool_id,omitempty" xor:"RedirectURL"`
```

Comment:
Just to confirm: `xor` will require _one_ of the two fields to be set. Is it possible to create an L7 policy without setting either of them? If so, then we need to remove the `xor`. Is it possible that the "reject" action will allow both to be unspecified (https://github.com/openstack/neutron-lbaas/blob/e4aba85427f696a6224b8ed6392e0ba16d6f98cf/neutron_lbaas/drivers/octavia/driver.py#L542) ?

These URLs might also be useful for how the Neutron logic works.

* https://github.com/openstack/neutron-lbaas/blob/eb5877fb004b06c912aa221225ea261773a48aa9/neutron_lbaas/db/loadbalancer/loadbalancer_dbv2.py#L429
* https://github.com/openstack/neutron-lbaas/blob/e4aba85427f696a6224b8ed6392e0ba16d6f98cf/neutron_lbaas/drivers/octavia/driver.py#L533

Reply:
you are right, these params don't need to be set if action is